### PR TITLE
feat: android Ti.UI.WebView support allowFileAccess

### LINKED ISSUE #13188: Android Webview android:targetSdkVersion=30+ access local files default changed
With Android SDK Version 30+ by default a webview will not have access to local assets like images accessed through file://.

This is therefore a breaking change since previously the default was for this setting to be true.

Android documentation of the change: https://developer.android.com/reference/android/webkit/WebSettings#setAllowFileAccess(boolean)

Stackoverflow dealing with the problem natively: https://stackoverflow.com/a/63846853/181197

I will submit a PR so that this becomes a property.

### PR DESCRIPTION
**Pull request by markive**
This is needed so that local files can be loaded into a WebView. 

In SDK 9.X.X this was allowed by default, in 10.X.X this is not the case. The Android documentation recommends that this is set to false by default so I have added this as a new Ti.UI.WebView property setAllowFileAccess() as per previous PR feedback

Fixes #13188

### CODE REVIEW

**Comment by hansemannn** [comment]
I feel like there was a reason for it to be removed. Kindly tagging @jquick-axway who most likely knows more here.

**Comment by jquick-axway** [comment]
This "feature" allows the website to use "file://" URLs to access local files. Apple and Google considers this to be a security issue. And it violates the "same origin" policy for URLs and exposes a cross-app/site scripting vulnerability. If the Android device is rooted, then it might expose the entire Android file system to the web view too.
https://developer.android.com/reference/android/webkit/WebSettings#setAllowFileAccess(boolean)

I'm pretty sure Google Play's security scanner will flag the app with a warning if it sees this API in the code. I'd be worried about all apps getting wrongly flagged.
https://stackoverflow.com/questions/50791306/google-play-warning-your-app-contains-a-file-based-xss-issue-javascript-enable
https://stackoverflow.com/questions/53095398/google-play-warning-your-app-contains-a-cross-app-scripting-vulnerability

On iOS, you have to use the `assetsDirectory` property and assign it the local directory or single file you want to expose to the `WebView`.
https://titaniumsdk.com/api/titanium/ui/webview.html#assetsdirectory

Maybe something similar can be implemented via the `WebViewAssetLoader`? This class' `shouldInterceptRequest()` method can intercept any URL (such as a `file://` URL) and return the file content the website is looking for... but only if the file "should" be exposed. Probably more work than what you guys want to implement.
https://developer.android.com/reference/android/webkit/WebResourceResponse.html

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/webview/TiUIWebView.java [hunk 1]
```diff
@@ -324,6 +324,13 @@ public TiUIWebView(TiViewProxy proxy)
 			}
 		}
 
+		boolean allowFileAccess = false; //file access should be false by default: https://developer.android.com/reference/android/webkit/WebSettings#setAllowFileAccess(boolean)
+		if (proxy.hasProperty(TiC.PROPERTY_ALLOW_FILE_ACCESS)) {
+			allowFileAccess = TiConvert.toBoolean(proxy.getProperty(TiC.PROPERTY_ALLOW_FILE_ACCESS), false);
+		}
+
+		settings.setAllowFileAccess(allowFileAccess);
+
```
Comment (m1ga) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/webview/TiUIWebView.java:333]: maybe change it too:
```
if (proxy.hasProperty(TiC.PROPERTY_ALLOW_FILE_ACCESS)) {
   settings.setAllowFileAccess(TiConvert.toBoolean(proxy.getProperty(TiC.PROPERTY_ALLOW_FILE_ACCESS), false));
}
```
so `setAllowFileAccess` will only be in the `if` part.

**Comment by m1ga** [comment]
There is an example for `WebViewAssetLoader` in https://developer.android.com/reference/androidx/webkit/WebViewAssetLoader

But it looks like we can use this PR too: https://github.com/tidev/titanium_mobile/issues/13188#issuecomment-1177675842 and it's not being flagged by Google.

**Comment by Xlinx64** [comment]
Can we merge it?

**Comment by caspahouzer** [comment]
We need @markive as a listed [contributor](https://tidev.io/contribute) to merge this

**Comment by markive** [comment]
@Xlinx64  @caspahouzer, sorry i'm not very Github experienced.. I think we are all good to merge, I've updated branch to merge latest changes.. Anything else I'm needed to do?

**Comment by caspahouzer** [comment]
@markive you have to sign the [contributor cla](https://tidev.io/contribute) as described. After you've been added to the [contributors list](https://github.com/tidev/organization-docs/blob/main/AUTHORIZED_CONTRIBUTORS.md) we can merge your changes.

**Comment by markive** [comment]
@caspahouzer ok I've reached out to Josh, thanks!

**Comment by markive** [comment]
@caspahouzer please note I have signed the CLA, so as soon as I appear on the contributor list this can be merged. Thanks!

**Comment by markive** [comment]
@caspahouzer @m1ga I am now on the contributor list so I think all is in place, thanks! https://github.com/tidev/organization-docs/blob/main/AUTHORIZED_CONTRIBUTORS.md

**Comment by Xlinx64** [comment]
Will this be part of the next release?
What is keeping us back to approve and merge this?

**Comment by markive** [comment]
Is there anything else I need to do? This is my first submission..

**Comment by m1ga** [comment]
@markive can you add it to the docs too. Have a look at https://github.com/tidev/titanium_mobile/blob/master/apidoc/Titanium/UI/WebView.yml#L727-L735 for an example. 

Also tried to test it but I wasn't successful. I've tried:
```js
const win = Ti.UI.createWindow({});
const webview = Ti.UI.createWebView({
	allowFileAccess: true,
});
win.add(webview);

win.addEventListener("open", function() {
	var permissions = [ 'android.permission.READ_EXTERNAL_STORAGE' ];
	Ti.Android.requestPermissions(permissions, function(e) {
		if (e.success) {
			webview.url = "file:////storage/emulated/0/test.html";
		}
	});
})
win.open();
```
but still get err_access_denied

**Comment by markive** [comment]
@m1ga I've retested it and it does work for me like this:

image

image

`const win = Ti.UI.createWindow({});
const webview = Ti.UI.createWebView({
	allowFileAccess: true,
});
win.add(webview);

win.addEventListener("open", function() {

            var assetFile = Ti.Filesystem.getFile(Titanium.Filesystem.resourcesDirectory, '/images/PlotIconBlue@2x.png');
            var destFile = Ti.Filesystem.getFile(Titanium.Filesystem.applicationDataDirectory, 'PlotIconBlue@2x.png')

            if (!destFile.exists()) {
                assetFile.copy(destFile.nativePath);
            }

			webview.url = "test.html";
})
win.open();`


`<!doctype html>
<html>
	<head>
		<title>View Plan</title>
		<meta name="viewport" content="width=device-width, maximum-scale=4, minimum-scale=0.1, user-scalable=yes">
		<meta name="format-detection" content="telephone=no"/>
		
		<style type="text/css">
			/* Prevent the default selection behaviour */
			html, body {
				-webkit-user-select: none;
				-webkit-touch-callout: none;
			}
		
			html, body, #plan-image { 
				margin: 0; padding: 0; 
			}
			
			#plan-image {
				position: relative;
				height: 1000px;
				width: 1000px;
				z-index: 1;		
			}
			
			.defect, .newDefect, .location, .drilldown, .mp { 				
				position: absolute; 
				cursor: pointer;
				z-index: 2;
				
				background-repeat: no-repeat;
				background-color: transparent;
				
				-webkit-background-size: 100% 100%;
          		background-size: 100% 100%;
			}
				
			.defect {
				width: 50px; 
				height: 50px;
				background-image: url(file:///android_asset/Resources/images/PlotIcon@2x.png);
			}

			.blueDefect {
				padding-left:10px;
				width: 50px; 
				height: 50px;
				background-image: url(file:///data/user/0/uk.co.marktestdelete/app_appdata/PlotIconBlue@2x.png);
			}
			
		</style>
	</head>
	<body>
		<div id="plan-image">
			<div class="defect" >&nbsp;</div>
			<div class="blueDefect" style="margin-left:10px">&nbsp;</div>
			<h1>hello!</h1>
		</div>
		
		
	</body>
</html>`

First test is with `allowFileAccess: false`, second is true.

**Comment by m1ga** [comment]
Thanks for the example @markive :+1: That helped and it's working fine. Testing it with Liveview wasn't a good idea since that cached the images in the webview even when you switch it off.

**Comment by hbugdoll** [comment]
PR can be closed in my opinion.

**Comment by hbugdoll** [comment]
@m1ga what is still needed?
Tests and checks have passed, @markive is listed as a authorized contributor.

**Comment by m1ga** [comment]
@hbugdoll I don't merge PRs, I'm only a contributor too. I have more than one PR waiting :wink: So we'll need to be patient (or you a custom SDK with this PR merged in)

**Comment by hbugdoll** [comment]
> @hbugdoll I don't merge PRs, I'm only a contributor too.

@m1ga Thank you for the clarification.

**Comment by m1ga** [comment]
To check if there is any automatic security scan that would flag an app I've uploaded an app with 12.1.0 (master) + this PR merge in. I didn't enable the flag so it's a "normal" app build like other users would do it too. 
No issues during app upload check and after the review. App is live, no warnings :+1:

**Comment by markive** [comment]
Yes I have also been using this in production across 2 different apps for over 1yr, so that is many many full reviews and approvals with no issue.

**Review by hansemannn — APPROVED** [review]

**Pull request merged**